packages/firestore/src/api/database.ts

```diff
@@ -753,7 +789,7 @@ export class DocumentReference<T = PublicDocumentData>
     return new DocumentReference<U>(
       this.firestore,
       this._delegate.withConverter(
-        converter as UntypedFirestoreDataConverter<U>
+        new UntypedFirestoreDataConverterAdapter(this.firestore, converter)
```

Comment:
Since we cannot force our users to implement a custom `isEqual` function, the problem you mentioned in chat can likely only be solved by looking at the type of data converter and "unwrapping" the adapter when needed.

This would look something like this:

```
const c1 = this.converter instanceof Compat ? this.converter._delegate : this.converter;
const c2 = that.converter instanceof Compat ? that.converter._delegate : that.converter;

if (c1 == c2) console.log("Woohoo!");
```

FWIW, a more straightforward fix may be to not call https://github.com/firebase/firebase-js-sdk/blob/566a62dd4d54ca6ca8bef8bb4409023050bb9f90/packages/firestore/src/exp/snapshot.ts#L264 when a converter is used, but to instead do something like this:


````
// In the compat class

data() {
  if (!this.converter) return this.delegate.data(); 
  else {
  const snapshot = new ExpQueryDocumentSnapshot(this._firestore, ...,
     new QueryDocumentSnapshot(
        this._firestore,
        this._userDataWriter,
        this._key,
        this._document,
        this.metadata,
        /* converter= */ null
      ));
      return this._converter.fromFirestore(snapshot, options);
  }
}


````

Reply:
I like this more "straightforward" fix. We would just have to be careful to implement `isEqual` to check equality of `this._converter` instead of relying on comparing the `_converter` of the delegate (since it will always be null). I think this is the best approach though so I'm going to try it out.

Reply:
I believe the `isEqual` problem goes away if we use the "straightforward fix.

Reply:
If we use the "straightforward fix" then the delegate's converter will always be null. Since `refEqual()` checks the converter for equality, that logic would need to be moved into the `isEqual()` methods to maintain the same semantics of `isEqual()`. Otherwise, the converters would not be considered when checking for equality, and the current logic *does* check the converters for equality.